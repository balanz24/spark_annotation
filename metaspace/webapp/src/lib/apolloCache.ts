import { defaultDataIdFromObject, InMemoryCache } from 'apollo-client-preset'

const nonNormalizableTypes: any[] = ['User', 'DatasetUser', 'DatasetGroup', 'DatasetProject']
export const makeApolloCache = () => new InMemoryCache({
  cacheRedirects: {
    Query: {
      // Allow get-by-id queries to use cached data that originated from other kinds of queries
      dataset: (_, args, { getCacheKey }) => getCacheKey({ __typename: 'Dataset', id: args.id }),
      annotation: (_, args, { getCacheKey }) => getCacheKey({ __typename: 'Annotation', id: args.id }),
      group: (_, args, { getCacheKey }) => getCacheKey({ __typename: 'Group', id: args.groupId }),
      project: (_, args, { getCacheKey }) => getCacheKey({ __typename: 'Project', id: args.projectId }),
    },
  },
  dataIdFromObject(object: any) {
    // WORKAROUND: Because of Apollo's aggressive caching, often the current User will be overwritten with results
    // from other queries. The server side often strips fields based on how they're accessed (the "ScopeRole" logic),
    // which means these query paths will often return different data with the same IDs:
    // currentUser -> primaryGroup (always present)
    // dataset -> submitter -> primaryGroup (null unless admin)
    // To protect against this, don't allow Users (and possibly other types in the future) to have a dataId,
    // so that InMemoryCache cannot share data between different queries.
    if (nonNormalizableTypes.includes(object.__typename)) {
      return null
    } else {
      return defaultDataIdFromObject(object)
    }
  },
})
