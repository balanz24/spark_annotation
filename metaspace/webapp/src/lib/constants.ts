export const MAX_MOL_DBS = 4
export const MAX_MOL_DBS_EXT = 6
export const MAX_NEUTRAL_LOSSES = 3
export const MAX_CHEM_MODS = 3
export const DEFAULT_SCALE_TYPE = 'linear'
