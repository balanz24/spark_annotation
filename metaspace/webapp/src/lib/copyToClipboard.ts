export default function copyToClipboard(text?: string) {
  if (text) {
    if ('clipboard' in navigator) {
      navigator.clipboard.writeText(text)
    } else {
      const el = document.createElement('textarea')
      el.value = text
      el.style.position = 'absolute'
      el.style.left = '-9999px'
      document.body.appendChild(el)
      try {
        el.select()
        document.execCommand('copy')
      } finally {
        document.body.removeChild(el)
      }
    }
  }
}
