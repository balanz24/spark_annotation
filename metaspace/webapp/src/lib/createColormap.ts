import * as d3 from 'd3'
import getColorScale from './getColorScale'

export type OpacityMode = 'constant' | 'linear' | 'quadratic';

const OPACITY_MAPPINGS: Record<OpacityMode, (x:number) => number> = {
  constant: (x) => 255,
  linear: (x) => x,
  quadratic: (x) => 255 - (255 - x) * (255 - x) / 255,
}

export default function createColormap(name: string, opacityMode: OpacityMode = 'constant',
  opacityScale: number = 1): number[][] {
  const { domain, range } = getColorScale(name)
  const sclFun = d3.scaleLinear<string>().domain(domain).range(range).clamp(true)
  const alphaFunc = OPACITY_MAPPINGS[opacityMode]

  const colors = []
  for (let i = 0; i < 256; i++) {
    const color = d3.rgb(sclFun(i / 255.0))
    const alpha = alphaFunc(i) * opacityScale
    colors.push([color.r, color.g, color.b, alpha].map(Math.round))
  }
  return colors
}
