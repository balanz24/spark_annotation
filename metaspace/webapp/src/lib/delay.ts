export default function delay(timeMs: number) {
  return new Promise(resolve => setTimeout(resolve, timeMs))
}
