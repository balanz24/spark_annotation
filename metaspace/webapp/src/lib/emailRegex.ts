// W3C's validation regex for <input type="email">. This is quite permissive, but doesn't allow unicode domain names
// (e.g. foo@mañana.com). To handle unicode domain names, the server-side code would have to be changed to automatically
// convert them to punycode when sending emails, as AWS SES doesn't support unicode domain names.

export default /^[a-zA-Z0-9.!#$%&’*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/
