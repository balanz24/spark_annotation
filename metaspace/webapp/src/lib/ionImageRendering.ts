import { decode, Image, toRGBA8 } from 'upng-js'
import { quantile } from 'simple-statistics'
import { range } from 'lodash-es'
import { DEFAULT_SCALE_TYPE } from './constants'
// @ts-ignore
import PxBrush from 'px-brush'

export interface IonImage {
  intensityValues: Float32Array;
  clippedValues: Uint8ClampedArray; // Intensity values scaled into 0-255
  mask: Uint8ClampedArray; // 0 = empty, 255 = filled, no other values
  width: number;
  height: number;
  minIntensity: number;
  maxIntensity: number;
  clippedMinIntensity: number;
  clippedMaxIntensity: number;
  scaledMinIntensity: number;
  scaledMaxIntensity: number;
  userMinIntensity: number;
  userMaxIntensity: number;
  // scaleBarValues - Quantization of linear intensity values, used for showing the distribution of colors on the scale bar
  // Always length 256
  scaleBarValues: Uint8ClampedArray;
  lowQuantile: number;
  highQuantile: number;
}

export interface Normalization {
  data: Float32Array | null
  shape: [number, number] | null,
  metadata: any,
  type: string | null,
  showFullTIC: boolean | null,
  error: boolean,
}

export type ColorMap = readonly number[][]

export type IonImageLayer = { ionImage: IonImage, colorMap: ColorMap }

export type ScaleType = 'linear' | 'linear-full' | 'log' | 'log-full' | 'hist' | 'test';
export type ScaleMode = 'linear' | 'log' | 'hist';

const SCALES: Record<ScaleType, [ScaleMode, number, number]> = {
  linear: ['linear', 0, 0.99],
  'linear-full': ['linear', 0, 1],
  log: ['log', 0.01, 0.99],
  'log-full': ['log', 0, 1],
  hist: ['hist', 0, 1],
  test: ['linear', 0, 0.5], // For unit tests, because it's easier to make test data for 50% threshold than 99%
}

const applyImageData = (canvas: HTMLCanvasElement, imageBytes: Uint8ClampedArray, width: number, height: number) => {
  if (imageBytes.length !== width * height * 4) {
    throw new Error('imageBytes must be in RGBA format')
  }
  canvas.width = width
  canvas.height = height
  const ctx = canvas.getContext('2d')!
  let imageData: ImageData
  try {
    imageData = new ImageData(imageBytes, width, height)
  } catch (ex) {
    // IE11 doesn't support `new ImageData`, so it gets the slow path
    imageData = ctx.createImageData(width, height)
    imageData.data.set(imageBytes)
  }
  ctx.putImageData(imageData, 0, 0)
}

const createDataUrl = (imageBytes: Uint8ClampedArray, width: number, height: number) => {
  const canvas = document.createElement('canvas')
  applyImageData(canvas, imageBytes, width, height)
  return canvas.toDataURL()
}

const getNormalizationIntensityAndMask = (normalizationData: Normalization) => {
  const intensityValues = normalizationData.data!
  const mask = new Uint8ClampedArray(intensityValues.length)
  for (let i = 0; i < intensityValues.length; i++) {
    mask[i] = Number.isNaN(intensityValues[i]) ? 0 : 255
  }
  const [width, height] = normalizationData.shape!

  return { intensityValues, mask, width, height }
}

const extractIntensityAndMask = (png: Image, min: number, max: number, normalizationData?: Normalization) => {
  const { width, height, depth, ctype } = png
  const hasAlpha = ctype === 4 || ctype === 6
  const isRGB = ctype === 6
  const bytesPerComponent = depth <= 8 ? 1 : 2
  const numPixels = (width * height)
  const numComponents = (ctype & 2 ? 3 : 1) + (hasAlpha ? 1 : 0)
  const rangeVal = Number(max - min) / (bytesPerComponent === 1 ? 255 : 65535)
  const baseVal = Number(min)
  // NOTE: pngDataBuffer usually has some trailing padding bytes. TypedArrays should have explicit sizes specified to
  // prevent over-reading // if hasAlpha, 16bit and not grayscale uses toRGBA to translate frames
  const pngDataBuffer = (hasAlpha && bytesPerComponent > 1 && !isRGB) ? toRGBA8(png)[0]
    : (png.data as any as Uint8Array).buffer // The typings are wrong
  const TIC_MULTIPLIER = 1000000

  // NOTE: This function is a bit verbose. It's intentionally structured this way so that the JS engine can
  // statically determine the types of all variables involved in copying, and hopefully generate fast machine code.
  const intensityValues = new Float32Array(numPixels)
  const mask = new Uint8ClampedArray(numPixels)
  const dataView = new DataView(pngDataBuffer, 0, numPixels * numComponents * bytesPerComponent)
  if (bytesPerComponent === 1) {
    for (let i = 0; i < numPixels; i++) {
      const byteOffset = i * numComponents * bytesPerComponent
      let intensity = dataView.getUint8(byteOffset) * rangeVal + baseVal
      // apply normalization
      if (
        normalizationData && normalizationData.data
        && normalizationData.data.length === numPixels
        && normalizationData.data[i] && !isNaN(normalizationData.data[i])) {
        intensity = (intensity / normalizationData.data[i]) * TIC_MULTIPLIER
      } else if (
        normalizationData && normalizationData.data
        && normalizationData.data.length === numPixels) {
        intensity = 0
      }

      intensityValues[i] = intensity
    }
    if (hasAlpha) {
      const alphaOffset = numComponents - 1
      for (let i = 0; i < numPixels; i++) {
        const byteOffset = i * numComponents * bytesPerComponent + alphaOffset
        mask[i] = dataView.getUint8(byteOffset) < 128 ? 0 : 255
      }
    } else {
      mask.fill(255)
    }
  } else {
    for (let i = 0; i < numPixels; i++) {
      const byteOffset = i * numComponents * bytesPerComponent
      let intensity = dataView.getUint16(byteOffset) * rangeVal + baseVal

      // apply normalization
      if (
        normalizationData && normalizationData.data
        && normalizationData.data.length === numPixels
        && normalizationData.data[i] && !isNaN(normalizationData.data[i])
      ) {
        intensity = (intensity / normalizationData.data[i]) * TIC_MULTIPLIER
      } else if (
        normalizationData && normalizationData.data
          && normalizationData.data.length === numPixels) {
        intensity = 0
      }
      intensityValues[i] = intensity
    }
    if (hasAlpha) {
      const alphaOffset = (numComponents - 1) * bytesPerComponent
      for (let i = 0; i < numPixels; i++) {
        const byteOffset = i * numComponents * bytesPerComponent + alphaOffset
        mask[i] = dataView.getUint16(byteOffset, !isRGB) < 32768 ? 0 : 255
      }
    } else {
      mask.fill(255)
    }
  }

  return { intensityValues, mask }
}

function safeQuantile(values: number[], q: number): number;
function safeQuantile(values: number[], q: number[]): number[];
function safeQuantile(values: number[], q: number | number[]): any {
  // Handle cases when `values` is empty gracefully
  if (values.length > 0) {
    return quantile(values, q as any)
  } else if (typeof q === 'number') {
    return 0
  } else {
    return q.map(() => 0)
  }
}

const getQuantileValues = (intensityValues: Float32Array, mask: Uint8ClampedArray, minValueConsidered: number = 0) => {
  const values = []
  for (let i = 0; i < mask.length; i++) {
    if (intensityValues[i] > minValueConsidered && mask[i] !== 0) {
      values.push(intensityValues[i])
    }
  }
  return values
}

const getRankValues = (values: number[], lowQuantile: number, highQuantile: number) => {
  const lo = lowQuantile || 0; const hi = highQuantile || 1
  const quantiles = range(256).map(i => lo + (hi - lo) * i / 255)
  return new Float32Array(safeQuantile(values, quantiles))
}

const quantizeIonImageLinear = (intensityValues: Float32Array, minIntensity: number, maxIntensity: number) => {
  const clippedValues = new Uint8ClampedArray(intensityValues.length)
  const intensityScale = 255 / (maxIntensity - minIntensity)
  for (let i = 0; i < intensityValues.length; i++) {
    clippedValues[i] = (intensityValues[i] - minIntensity) * intensityScale
  }

  return clippedValues
}

const quantizeIonImageLog = (intensityValues: Float32Array, minIntensity: number, maxIntensity: number) => {
  const clippedValues = new Uint8ClampedArray(intensityValues.length)
  const logMinIntensity = Math.log(minIntensity)
  const intensityScale = 255 / (Math.log(maxIntensity) - logMinIntensity)

  for (let i = 0; i < intensityValues.length; i++) {
    clippedValues[i] = (Math.log(intensityValues[i]) - logMinIntensity) * intensityScale
  }

  return clippedValues
}

const quantizeIonImageRank = (intensityValues: Float32Array, rankValues: Float32Array) => {
  const clippedValues = new Uint8ClampedArray(intensityValues.length)
  const min = rankValues[0]
  const max = rankValues[255]

  for (let i = 0; i < intensityValues.length; i++) {
    const intensity = intensityValues[i]
    if (intensity < min) {
      clippedValues[i] = 0
    } else if (intensity >= max) {
      clippedValues[i] = 255
    } else {
      // Fixed-depth binary search into 256-element array
      let mid = 0
      if (intensity > rankValues[mid | 0x80]) mid |= 0x80
      if (intensity > rankValues[mid | 0x40]) mid |= 0x40
      if (intensity > rankValues[mid | 0x20]) mid |= 0x20
      if (intensity > rankValues[mid | 0x10]) mid |= 0x10
      if (intensity > rankValues[mid | 0x08]) mid |= 0x08
      if (intensity > rankValues[mid | 0x04]) mid |= 0x04
      if (intensity > rankValues[mid | 0x02]) mid |= 0x02
      if (intensity > rankValues[mid | 0x01]) mid |= 0x01
      clippedValues[i] = mid
    }
  }

  return clippedValues
}

const quantizeIonImage = (intensityValues: Float32Array, minIntensity: number, maxIntensity: number,
  rankValues: Float32Array | null, scaleMode: ScaleMode): Uint8ClampedArray => {
  if (scaleMode === 'hist') {
    return quantizeIonImageRank(intensityValues, rankValues!)
  } else if (scaleMode === 'log') {
    return quantizeIonImageLog(intensityValues, minIntensity, maxIntensity)
  } else {
    return quantizeIonImageLinear(intensityValues, minIntensity, maxIntensity)
  }
}

const quantizeScaleBar = (minIntensity: number, maxIntensity: number,
  rankValues: Float32Array | null, scaleMode: ScaleMode): Uint8ClampedArray => {
  const linearDistribution =
    new Float32Array(range(256).map(i => minIntensity + (maxIntensity - minIntensity) * i / 255))
  return quantizeIonImage(linearDistribution, minIntensity, maxIntensity, rankValues, scaleMode)
}

export const loadPngFromUrl = async(url: string): Promise<Image> => {
  const response = await fetch(url, { credentials: 'omit' })
  if (response.status !== 200) {
    throw Object.assign(
      new Error(`Invalid response fetching image: ${response.status} ${response.statusText}`),
      // Don't report 403s/404s to Sentry - they're virtually always caused by deleted datasets
      { isHandled: response.status === 403 || response.status === 404 },
    )
  }
  const buffer = await response.arrayBuffer()
  return decode(buffer)
}

export const processIonImage = (
  png: Image, minIntensity: number = 0, maxIntensity: number = 1, scaleType: ScaleType = DEFAULT_SCALE_TYPE,
  userScaling: readonly [number, number] = [0, 1],
  userIntensities: readonly [number?, number?] = [],
  normalizationData?: Normalization): IonImage => {
  const [scaleMode, lowQuantile, highQuantile] = SCALES[scaleType]
  const { width, height } = png
  let [userMin = minIntensity, userMax = maxIntensity] = userIntensities

  const { intensityValues, mask } = normalizationData?.showFullTIC
    ? getNormalizationIntensityAndMask(normalizationData)
    : extractIntensityAndMask(png, minIntensity, maxIntensity, normalizationData)

  // assign normalized intensities
  if (normalizationData && normalizationData.metadata && scaleType !== 'hist') {
    maxIntensity = normalizationData ? maxIntensity / normalizationData.metadata.maxTic * 1000000 : maxIntensity
    userMax = maxIntensity
  }

  // Only non-zero values should be considered for hotspot removal, otherwise sparse images have most of their set pixels treated as hotspots.
  // For compatibility with the previous version where images were loaded as 8-bit, linear scale's thresholds exclude pixels
  // whose values would round down to zero. This can make a big difference - some ion images have as high as 40% of
  // their pixels set to values that are zero when loaded as 8-bit but non-zero when loaded as 16-bit.
  const minValueConsidered = (scaleMode === 'linear' ? maxIntensity / 256 : 0)
  const quantileValues = getQuantileValues(intensityValues, mask, minValueConsidered)

  let min = minIntensity
  if (userMin !== minIntensity) {
    min = userMin
  } else if (scaleMode === 'log' || lowQuantile > 0) {
    min = safeQuantile(quantileValues, lowQuantile)
  }
  let max = maxIntensity
  if (userMax !== maxIntensity) {
    max = userMax
  } else if (highQuantile < 1) {
    max = safeQuantile(quantileValues, highQuantile)
  }

  const [minScale, maxScale] = userScaling
  let scaledMin = min + ((max - min) * minScale)
  let scaledMax = min + ((max - min) * maxScale)

  let rankValues = null
  if (scaleType === 'hist') {
    const { intensityValues } = normalizationData?.showFullTIC
      ? getNormalizationIntensityAndMask(normalizationData)
      : extractIntensityAndMask(png, scaledMin, scaledMax, normalizationData)
    const values = getQuantileValues(intensityValues, mask)
    rankValues = getRankValues(values, lowQuantile, highQuantile)

    // reassign intensity values to be displayed if normalized
    if (normalizationData && normalizationData.metadata) {
      maxIntensity = normalizationData ? maxIntensity / normalizationData.metadata.maxTic * 1000000 : maxIntensity
      userMax = maxIntensity
      max = maxIntensity

      if (userMax !== maxIntensity) {
        max = userMax
      } else if (highQuantile < 1) {
        max = safeQuantile(quantileValues, highQuantile)
      }
      scaledMin = min + ((max - min) * minScale)
      scaledMax = min + ((max - min) * maxScale)
    }
  }

  const clippedValues = quantizeIonImage(intensityValues, scaledMin, scaledMax, rankValues, scaleMode)
  const scaleBarValues = quantizeScaleBar(scaledMin, scaledMax, rankValues, scaleMode)

  return {
    intensityValues,
    clippedValues,
    mask,
    width,
    height,
    minIntensity,
    maxIntensity,
    clippedMinIntensity: min,
    clippedMaxIntensity: max,
    scaledMinIntensity: scaledMin,
    scaledMaxIntensity: scaledMax,
    userMinIntensity: userMin,
    userMaxIntensity: userMax,
    scaleBarValues,
    lowQuantile,
    highQuantile,
  }
}

function getCmapComponents(cmap: ColorMap, buffer : ArrayBuffer = new ArrayBuffer(256 * 4)) {
  const cmapComponents = new Uint8ClampedArray(buffer)
  for (let i = 0; i < 256; i++) {
    if (cmap != null) {
      for (let c = 0; c < 4; c++) {
        cmapComponents[i * 4 + c] = cmap[i][c]
      }
    } else {
      for (let c = 0; c < 4; c++) {
        cmapComponents[i * 4 + c] = i
      }
    }
  }
  return cmapComponents
}

export const renderIonImageToBuffer = (ionImage: IonImage, cmap: readonly number[][], buffer?: ArrayBuffer) => {
  const { clippedValues, mask } = ionImage
  // Treat pixels as 32-bit values instead of four 8-bit values to avoid extra math.
  // Assume little-endian byte order, because big-endian is pretty much gone.
  const outputBuffer = buffer || new ArrayBuffer(clippedValues.length * 4)
  const outputRGBA = new Uint32Array(outputBuffer)
  const cmapBuffer = new ArrayBuffer(256 * 4)
  const cmapRGBA = new Uint32Array(cmapBuffer)
  const emptyRGBA = 0x00000000

  getCmapComponents(cmap, cmapBuffer)

  for (let i = 0; i < mask.length; i++) {
    if (mask[i]) {
      outputRGBA[i] += cmapRGBA[clippedValues[i]]
    } else {
      outputRGBA[i] += emptyRGBA
    }
  }
  return outputBuffer
}

export const renderIonImages = (layers: IonImageLayer[], canvas: HTMLCanvasElement, width: number, height: number,
  roiInfo?: any[]) => {
  const ctx = canvas.getContext('2d')!
  ctx.clearRect(0, 0, width, height)
  const pxBrush = new PxBrush(canvas)

  if (layers.length === 0) return

  const [base] = layers
  const { clippedValues } = base.ionImage

  const buffer = new ArrayBuffer(clippedValues.length * 4)
  const pixels = new Uint8ClampedArray(buffer)

  for (const { ionImage, colorMap } of layers) {
    const cmapComponents = getCmapComponents(colorMap)
    const { clippedValues, mask } = ionImage

    for (let i = 0; i < mask.length; i++) {
      if (!mask[i]) {
        pixels[i * 4] = 0
        pixels[i * 4 + 1] = 0
        pixels[i * 4 + 2] = 0
        pixels[i * 4 + 3] = 0
      } else {
        const v = clippedValues[i]

        for (let j = 0; j < 3; j++) {
          pixels[i * 4 + j] += cmapComponents[v * 4 + j]
        }

        // inspired by https://github.com/colorjs/color-composite/blob/master/index.js#L30
        const a1 = cmapComponents[v * 4 + 3] / 255
        const a2 = pixels[i * 4 + 3] / 255
        pixels[i * 4 + 3] = (a1 + a2 * (1 - a1)) * 255
      }
    }
    applyImageData(canvas, pixels, width, height)

    // ROI drawing
    if (roiInfo && roiInfo.length > 0 && roiInfo[roiInfo.length - 1].allVisible) {
      roiInfo.forEach((roiItem: any) => {
        if (roiItem.visible && roiItem.coordinates.length > 0) {
          ctx.beginPath()
          ctx.strokeStyle = roiItem.strokeColor
          ctx.fillStyle = roiItem.color
          ctx.lineWidth = 1
          ctx.moveTo(roiItem.coordinates[0].x, roiItem.coordinates[0].y)

          for (let index = 1; index < roiItem.coordinates.length; index++) {
            pxBrush.brush({
              from: { x: roiItem.coordinates[index - 1].x, y: roiItem.coordinates[index - 1].y },
              to: { x: roiItem.coordinates[index].x, y: roiItem.coordinates[index].y },
              size: 1,
              color: roiItem.rgb,
            })
            ctx.lineTo(roiItem.coordinates[index].x + 0.5, roiItem.coordinates[index].y + 0.5)
            if (roiItem.coordinates[index].isEndPoint) { // closes path if last point
              ctx.closePath()
            }
          }
          ctx.fill()
          ctx.stroke()
        }
      })
    }
  }
}

export const renderScaleBar = (ionImage: IonImage, cmap: ColorMap, horizontal: boolean) => {
  const outputBytes = new Uint8ClampedArray(256 * 4)
  for (let i = 0; i < ionImage.scaleBarValues.length; i++) {
    const val = ionImage.scaleBarValues[i]
    for (let j = 0; j < 4; j++) {
      outputBytes[(horizontal ? i : 255 - i) * 4 + j] = cmap[val][j]
    }
  }

  if (horizontal) {
    return createDataUrl(outputBytes, 256, 1)
  } else {
    return createDataUrl(outputBytes, 1, 256)
  }
}

export const getIonImage = (ionImagePng: any, isotopeImage: any,
  scaleType: any = 'linear', userScaling: any = [0, 1], normalizedData: any = null) => {
  if (!isotopeImage || !ionImagePng) {
    return null
  }
  const { minIntensity, maxIntensity } = isotopeImage
  return processIonImage(ionImagePng, minIntensity, maxIntensity, scaleType
    , userScaling, undefined, normalizedData)
}
