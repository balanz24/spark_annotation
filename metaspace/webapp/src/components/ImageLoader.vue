<template>
  <div
    ref="container"
    v-resize="onResize"
  >
    <ion-image-viewer
      ref="imageLoader"
      :ion-image-layers="ionImageLayers"
      :is-loading="ionImageIsLoading"
      :width="imageFit.areaWidth"
      :height="imageFit.areaHeight"
      :pixel-aspect-ratio="pixelAspectRatio"
      :zoom="zoom"
      :x-offset="xOffset"
      :y-offset="yOffset"
      :style="imageStyle"
      v-bind="$attrs"
      v-on="$listeners"
    />
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import resize from 'vue-resize-directive'
import config from '../lib/config'
import { Component, Prop, Watch } from 'vue-property-decorator'
import IonImageViewer from './IonImageViewer'
import { IonImage, loadPngFromUrl, processIonImage, ScaleType, IonImageLayer } from '../lib/ionImageRendering'
import fitImageToArea, { FitImageToAreaResult } from '../lib/fitImageToArea'
import reportError from '../lib/reportError'
import createColormap, { OpacityMode } from '../lib/createColormap'

  @Component({
    inheritAttrs: false,
    directives: {
      resize,
    },
    components: {
      IonImageViewer,
    },
  })
export default class ImageLoader extends Vue {
    $refs: any;

    @Prop()
    src!: string | null;

    @Prop()
    imagePosition!: any;

    @Prop()
    imageStyle!: any;

    @Prop()
    minIntensity?: number;

    @Prop()
    maxIntensity?: number;

    @Prop()
    pixelAspectRatio!: number;

    @Prop({ type: String })
    scaleType?: ScaleType;

    @Prop({ default: 'Viridis' })
    colormap!: string;

    @Prop()
    opacityMode?: OpacityMode;

    @Prop()
    annotImageOpacity?: number;

    @Prop()
    normalizationData?: any;

    containerWidth = 500;
    containerHeight = 500;
    ionImage: IonImage | null = null;
    ionImageIsLoading = false;

    created() {
      const ignoredPromise = this.updateIonImage()
    }

    mounted() {
      this.onResize()
    }

    onResize() {
      if (this.$refs.container != null) {
        this.containerWidth = this.$refs.container.clientWidth
        this.containerHeight = this.$refs.container.clientHeight
      }
    }

    async updateIonImage() {
      // Keep track of which image is loading so that this can bail if src changes before the download finishes
      const newUrl = this.src

      if (newUrl != null) {
        this.ionImageIsLoading = true
        try {
          const png = await loadPngFromUrl((config.imageStorage || '') + newUrl)

          if (newUrl === this.src) {
            this.ionImage = processIonImage(png, this.minIntensity, this.maxIntensity, this.scaleType,
              undefined, undefined, this.normalizationData)
            this.ionImageIsLoading = false
          }
        } catch (err) {
          reportError(err, null)
          if (newUrl === this.src) {
            this.ionImage = null
            this.ionImageIsLoading = false
          }
        }
      }
    }

    @Watch('src')
    async updateImageSrc() {
      this.updateIonImage()
    }

    @Watch('normalizationData')
    async updateIonImageNormalization() {
      this.updateIonImage()
    }

    get zoom() {
      return (this.imagePosition && this.imagePosition.zoom || 1) * this.imageFit.imageZoom
    }

    get xOffset() {
      return this.imagePosition && this.imagePosition.xOffset || 0
    }

    get yOffset() {
      return this.imagePosition && this.imagePosition.yOffset || 0
    }

    get imageFit(): FitImageToAreaResult {
      return fitImageToArea({
        imageWidth: this.ionImage ? this.ionImage.width : this.containerWidth,
        imageHeight: (this.ionImage ? this.ionImage.height : this.containerHeight) / this.pixelAspectRatio,
        areaWidth: this.containerWidth,
        areaHeight: this.containerHeight,
      })
    }

    get ionImageLayers(): IonImageLayer[] {
      if (this.ionImage) {
        return [{
          ionImage: this.ionImage,
          colorMap: createColormap(this.colormap, this.opacityMode, this.annotImageOpacity),
        }]
      }
      return []
    }

    @Watch('imageFit')
    emitRedrawEvent() {
      this.$emit('redraw', {
        width: this.imageFit.areaWidth,
        height: this.imageFit.areaHeight,
        naturalWidth: this.ionImage ? this.ionImage.width : 0,
        naturalHeight: this.ionImage ? this.ionImage.height : 0,
      })
    }
}
</script>
