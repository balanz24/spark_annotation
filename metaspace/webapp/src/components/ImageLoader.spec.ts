import { mount } from '@vue/test-utils'
import Vue from 'vue'
import ImageLoader from './ImageLoader.vue'
import * as _ionImageRendering from '../lib/ionImageRendering'
import { IonImage } from '../lib/ionImageRendering'
import { range } from 'lodash-es'

jest.mock('../lib/ionImageRendering')
const mockIonImageRendering = _ionImageRendering as jest.Mocked<typeof _ionImageRendering>

const W = 200
const H = 300

describe('ImageLoader', () => {
  const baseProps = {
    src: 'http://placebacon.com/200/300',
    imagePosition: { zoom: 1, xOffset: 0, yOffset: 0 },
    imageStyle: { transform: 'scaleX(1.23)' },
    maxIntensity: 255,
    minIntensity: 0,
    pixelAspectRatio: 1,
    scaleType: undefined,
    colormap: 'Hot',
    imageFitParams: { areaHeight: 250, areaMinHeight: 50 },
    showPixelIntensity: true,
  }
  const fullProps = {
    ...baseProps,
    opticalSrc: 'http://placekitten.com/200/300',
    opacityMode: 'constant',
    opticalTransform: [[1, 2, 3], [4, 5, 6], [7, 8, 9]],
    annotImageOpacity: 0.8,
    scrollBlock: true,
    showPixelIntensity: true,
  }
  mockIonImageRendering.loadPngFromUrl.mockImplementation((url) => ({ url } as any))
  mockIonImageRendering.processIonImage.mockImplementation((png, minIntensity, maxIntensity, scaleType): IonImage => ({
    png,
    minIntensity,
    maxIntensity,
    scaleType,
    width: W,
    height: H,
    mask: new Uint8ClampedArray(new Array(W * H).fill(255)),
    intensityValues: new Float32Array(range(W * H)),
  } as any))
  mockIonImageRendering.renderIonImages.mockImplementation(
    (layers: any, canvas: any) => {
      if (canvas) {
        canvas.setAttribute('data-images', layers.map((_: any) => _.ionImage.png.url).join(', '))
      }
      return undefined
    },
  )

  beforeEach(() => {
    // Set HTMLElements to have non-zero dimensions
    // @ts-ignore
    jest.spyOn(HTMLElement.prototype, 'getBoundingClientRect').mockImplementation(() =>
      ({ left: 200, right: 200 + W, top: 100, bottom: 100 + H, width: W, height: H }))
    jest.spyOn(HTMLElement.prototype, 'clientWidth', 'get').mockImplementation(() => W)
    jest.spyOn(HTMLElement.prototype, 'clientHeight', 'get').mockImplementation(() => H)
  })

  it('should match snapshot (minimal)', async() => {
    const wrapper = mount(ImageLoader, { propsData: baseProps })
    await Vue.nextTick()

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should match snapshot (with everything turned on)', async() => {
    const wrapper = mount(ImageLoader, { propsData: fullProps })
    await Vue.nextTick()

    // Trigger mouseover to show the intensity popup.
    wrapper.find('div>div').trigger('mousemove', {
      clientX: 250,
      clientY: 150,
    })
    await Vue.nextTick()

    expect(wrapper.element).toMatchSnapshot()
  })
})
