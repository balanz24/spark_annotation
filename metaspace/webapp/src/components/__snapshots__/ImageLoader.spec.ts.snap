// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageLoader should match snapshot (minimal) 1`] = `
<div>
  <div
    class="relative overflow-hidden"
    imagefitparams="[object Object]"
    mock-v-loading="false"
    style="width: 200px; height: 300px; transform: scaleX(1.23);"
  >
    <div
      data-test-key="ion-image-panel"
      style="left: 0px; top: 0px; width: 200px; height: 300px; transform-origin: 0 0; transform: translate(0px, 0px) scale(1, 1);"
    >
      <canvas
        class="absolute top-0 left-0 z-10 origin-top-left select-none pixelated"
        data-images="http://placebacon.com/200/300"
        height="300"
        width="200"
      />
      <div />
    </div>
  </div>
</div>
`;

exports[`ImageLoader should match snapshot (with everything turned on) 1`] = `
<div>
  <div
    class="relative overflow-hidden"
    imagefitparams="[object Object]"
    mock-v-loading="false"
    style="width: 200px; height: 300px; transform: scaleX(1.23);"
  >
    <div
      class=""
      data-test-key="ion-image-panel"
      style="left: 0px; top: 0px; width: 200px; height: 300px; transform-origin: 0 0; transform: translate(0px, 0px) scale(1, 1);"
    >
      <canvas
        class="absolute top-0 left-0 z-10 origin-top-left select-none pixelated"
        data-images="http://placebacon.com/200/300"
        height="300"
        width="200"
      />
      <div>
        <img
          class="absolute top-0 left-0 -z-10 origin-top-left"
          crossorigin="anonymous"
          src="http://placekitten.com/200/300"
          style="transform: matrix3d(1, 4, 0, 7,
             2, 5, 0, 8,
                      0,          0, 1,          0,
             3, 6, 0, 9);"
        />
      </div>
    </div>
    <div>
      <mock-el-tooltip
        manual="true"
        placement="top"
        popper-class="pointer-events-none"
        value="true"
      >
        <div
          slot-key="content"
        >
          <div
            class="flex flex-col"
          >
            <div>
              <span>
                9.6e+3
              </span>
            </div>
          </div>
        </div>
        <div
          slot-key="default"
        >
          <div
            class="absolute block border-solid z-30 pointer-events-none box-border"
            style="left: 47.5px; top: 47.5px; width: 0.5px; height: 0.5px;"
          />
        </div>
      </mock-el-tooltip>
    </div>
    <div
      class="absolute inset-0 z-30 pointer-events-none bg-body flex items-center justify-center opacity-0 duration-1000"
    >
      <p
        class="relative block z-40 m-0 text-white text-2xl"
      >
        Use CTRL + scroll wheel to zoom the image
      </p>
    </div>
  </div>
</div>
`;
